platform/source/minimal-printf/mbed_printf_implementation.c

```diff
@@ -130,106 +139,103 @@ static void mbed_minimal_putchar(char *buffer, size_t length, int *result, char
+                    scratch[index] = int2hex_lower[output];
+                } else {
+                    scratch[index] = int2hex_upper[output];
+                }
 
-            /* shift value one decimal position */
-            value = value / 10;
-        }
+                /* shift value one byte position */
+                value = value >> 4;
+            } else {
+                /* use '0' as base and add digit */
+                scratch[index] = '0' + (value % 10);
 
-        /* write scratch pad to buffer or output */
-        for (; index > 0; index--) {
-            mbed_minimal_putchar(buffer, length, result, scratch[index - 1], stream);
+                /* shift value one decimal position */
+                value = value / 10;
+            }
         }
     }
-}
 
-/**
- * @brief      Print hexadecimal.
- *
- * @param      buffer  The buffer to store output (NULL for stdout).
- * @param[in]  length  The length of the buffer.
- * @param      result  The current output location.
- * @param[in]  value   The value to be printed.
- * @param      upper   Flag to print the hexadecimal in upper or lower case.
- */
-static void mbed_minimal_formatted_string_hexadecimal(char *buffer, size_t length, int *result, MBED_UNSIGNED_STORAGE value, FILE *stream, bool upper)
-{
-    bool print_leading_zero = false;
-
-    for (int index = 7; index >= 0; index--) {
-        /* get most significant byte */
-        uint8_t output = value >> (8 * index);
-
-        /* only print leading zeros when set */
-        if (print_leading_zero || (output != 0) || (index == 0)) {
-            unsigned int nibble_one = (output >> 4);
-            unsigned int nibble_two = (output & 0x0F);
-
-            static const char int2hex_lower[16] = { '0', '1', '2', '3', '4', '5', '6', '7',
-                                                    '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'
-                                                  };
-            static const char int2hex_upper[16] = { '0', '1', '2', '3', '4', '5', '6', '7',
-                                                    '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'
-                                                  };
-            const char *int2hex = upper ? int2hex_upper : int2hex_lower;
-
-            if (print_leading_zero || nibble_one != 0) {
-                mbed_minimal_putchar(buffer, length, result, int2hex[nibble_one], stream);
-            }
-            mbed_minimal_putchar(buffer, length, result, int2hex[nibble_two], stream);
+    if (negativ_value && !prepend_zeros) {
+        scratch[index] = '-';
+        index++;
+    }
+
+    if (index < width_size) {
+        index = width_size;
+    }
 
-            /* print zeroes after the first non-zero byte */
-            print_leading_zero = true;
+    if (negativ_value && prepend_zeros) {
+        if (scratch[index - 1] == '0') {
+            scratch[index - 1] = '-';
+        } else {
+            scratch[index] = '-';
+            index++;
```

Comment:
Some comments here would make it easier to understand:
* In the `if` case, the actual width has not reached the minimum width, and the `-` sign is _included_ in the minimum width
* In the `else` case, the minimum width has been reached.